File: cli/src/main/java/io/microconfig/cli/commands/ShowCommand.java
package io.microconfig.cli.commands;

import java.net.URI;
import java.util.Optional;

import static io.microconfig.cli.util.HttpUtil.httpGET;
import static io.microconfig.cli.util.HttpUtil.httpSend;

public class ShowCommand extends Command {

    public ShowCommand(String[] args) {
        super(args);
    }

    @Override
    public int execute() {
        var component = component(helpMessage());

        var type = flags.type();
        var env = flags.env();
        var uri = uri(
                type.orElse("app"),
                component,
                env.orElse("default")
        );

        var request = httpGET(uri, flags.timeout());
        addFlags(request);
        var body = httpSend(request.build());
        System.out.println(body);
        return 0;
    }

    private URI uri(String type, String name, String env) {
        return URI.create(String.format("%s/api/config/%s/%s/%s", server(), type, name, env));
    }

    private String helpMessage() {
        return "Usage microctl show [component] [flags]\n"
                + "Generates configuration for component of specified type and outputs it to console\n"
                + "Flags: \n"
                + "  -e, --env:   config environment\n"
                + "  -t, --type:  config type, 'app' by default\n";
    }

}

File: cli/src/main/kotlin/io/microconfig/cli/commands/ShowCommand.kt
package io.microconfig.cli.commands

import io.microconfig.cli.CliException

class ShowCommand(args: Array<String>) : Command(args) {

    override fun execute(): Int {
        val component = component(helpMessage())
        val type = flags.type() ?: "app"

        configs(component, type)
            .firstOrNull { it.type == type }
            ?.let { println(it.content) }
            ?: throw CliException("No component with requested type", 404)

        return 0
    }

    private fun helpMessage(): String {
        return """
            Usage microctl show [component] [flags]
            Generates configuration for component of specified type and outputs it to console
            Flags: 
              -e, --env  [name]: config environment
              -t, --type [name]: config type, 'app' by default
              -s, --set  [foo=bar]: override values for placeholders and vars
            """.trimIndent()
    }
}